### ISSUE TITLE
Do a binary search to find the best wheel lazily?

### ISSUE THREAD

**Issue by wolfv** [prefix-dev/rip#95]
Sometimes we end up downloading loads of wheels. I would need to find a test case but I think `boto` sometimes ends up doing that.

In this case we might be better off by doing a binary search for the best matching version by more intelligently finding the first fitting one.

**Comment by notatallshaw** [comment]
This should be a good, but very simple, set of requirements to test against:

```
"urllib3[socks]<3" "boto3<1.28.57"
```

**Comment by notatallshaw** [comment]
I'm sure you're aware but bear in mind, that you could have Package A versions 1, 2, 3, 4, ..., 99, 100, and only Package A versions 33, 50, 51, and 80 are compatible in your entire dependency graph.

**Comment by wolfv** [comment]
Do you have a (large) wheel cache with pip? Or did you notice any other "strategy" that pip uses to get a faster resolution speed?

**Comment by notatallshaw** [comment]
> Do you have a (large) wheel cache with pip? Or did you notice any other "strategy" that pip uses to get a faster resolution speed?

No, the cache is unrelated, pip is visting a much smaller number of candidates before solving this requirement. Pip has implemented a few different optimizations that may be helping here, most recently extras are used to constrain the resolution path https://github.com/pypa/pip/pull/12095, backjumping was implemented earlier in the year https://github.com/sarugaku/resolvelib/pull/113, and prefering causes was implemented some time ago pip https://github.com/pypa/pip/pull/10481, but it would take a long time to figure out what is contributing.

But I also think that sometimes any given resolution algorithm can just be unlucky and take the wrong path, I created this example from when Pip was slow with these requirements: https://github.com/pypa/pip/issues/12274

Additionally though it does seem rip is backtracking fairly slowly through old versions of boto3 and botocore, when I first ran it it looked like it had to compile a bunch of old sdists and was taking 2 seconds per step. Running it a second time I guess that dependency information is cached(?), it still seems to take about 1/3 of a second per step. But I am not familiar with debugging rust code so this is just observations from looking at the logs.

**Comment by lambdaq** [comment]
I am new to this but has anyone tried to  parse wheels without full download?

1.  assume wheels are in ZIP format
2. there's metadata at the end of ZIP file.
3. use [Http Range](https://github.com/saulpw/unzip-http) requests to get the `pyproject.toml` or `requirements.txt`, inside the wheel

This could save lots of network time?

Also I noticed the pip library has a total size of 13MB. If ppl find this acceptable, this should also be enough to store every deps on pypi locally 🤣

**Comment by wolfv** [comment]
Yes, that's how we do it! :)
